File: 67. Add Binary/67.java
// 67. Add Binary solution in Java
// https://leetcode.com/problems/add-binary/description/

class Solution {
    public String addBinary(String a, String b) {
        StringBuilder result = new StringBuilder();

        int i = a.length() - 1;
        int j = b.length() - 1;
        int carry = 0;

        while (i >= 0 || j >= 0 || carry > 0) {
            char char_a;
            char char_b;

            if (i >= 0) {
                char_a = a.charAt(i);
            } else {
                char_a = '0';
            }

            if (j >= 0) {
                char_b = b.charAt(j);
            } else {
                char_b = '0';
            }

            int sum = Character.getNumericValue(char_a) + Character.getNumericValue(char_b) + carry;

            if (sum > 1) {
                sum = sum % 2;
                carry = 1;
            } else {
                carry = 0;
            }

            result.insert(0, sum);
            i--;
            j--;
        }

        return result.toString();
    }
}

File: 67. Add Binary/67.go
// 67. Add Binary solution in GoLang
// https://leetcode.com/problems/add-binary/description/

package leetcode

func addBinary(a string, b string) string {
    result := ""
    i := len(a) - 1
    j := len(b) - 1
    carry := 0

    for i >= 0 || j >= 0 {
        sum := carry
        
        if i >= 0 && string(a[i]) == "1" {
            sum += 1
        }

        if j >= 0 && string(b[j]) == "1" {
            sum += 1
        }

        if sum % 2 == 1 {
            result = "1" + result
        } else {
            result = "0" + result
        }

        if sum < 2 {
            carry = 0
        } else {
            carry = 1
        }

        i--
        j--
    }

    if carry != 0 {
        result = "1" + result
    }

    return result
}